test/lib/rules/no-this-property-during-ssr.js

```diff
@@ -0,0 +1,590 @@
+                    // we can't use this.template here
+                  }
+                  renderedCallback() {
+                    this.template.querySelector('span')?.foo();
+                  }
+                  bar() {
+                    this.template.querySelector('span')?.foo();
+                  }
+                }
+            `,
+        },
+        {
+            code: `
+                import { LightningElement } from 'lwc';
+                import tmplA from './a.html';
+
+                export default class Foo extends LightningElement {
+                  connectedCallback() {
+                    // we can't use this.template here
+                  }
+                  renderedCallback() {
+                    this.bar();
+                  }
+                  bar() {
+                    this.template.querySelector('span')?.foo();
+                  }
+                }
+            `,
+        },
+        {
+            code: `
+                import { LightningElement } from 'lwc';
+                import tmplA from './a.html';
+
+                export default class Foo extends LightningElement {
+                  connectedCallback() {
+                    // we can't use this.template here
+                  }
+                  renderedCallback() {
+                    this.bar();
+                  }
+                  bar() {
+                    doSomething(this.template);
+                  }
+                }
+            `,
+        },
+        {
+            code: `
+                import { LightningElement } from 'lwc';
+                import tmplA from './a.html';
+
+                export default class Foo extends LightningElement {
+                  connectedCallback() {
+                    // we can't use this.template here
+                  }
+                  bar() {
+                    doSomething(this.template);
+                  }
+                }
+            `,
+        },
+        {
+            code: `
+                import { LightningElement } from 'lwc';
+
+                export default class Foo extends LightningElement {
+                  connectedCallback() {
+                    if(document !== undefined) {
+                      doSomething(this.template);
+                    }
+                  }
+                }
+            `,
+        },
+    ],
+    invalid: [
+        {
+            code: `
+                import { LightningElement } from 'lwc';
+                import tmplA from './a.html';
+
+                export default class Foo extends LightningElement {
+                  connectedCallback() {
+                    this.template.querySelector('span')?.foo();
+                  }
+                }
+            `,
+            errors: [
+                {
+                    message:
+                        'You should not use `this.template.querySelector` in methods that will execute during SSR. Use `lwc:ref` instead.',
```

Comment:
This error message is misleading. Even with refs, you can't access the rendered element in the `connectedCallback` (on the client nor the server).